climada/hazard/tc_surge_geoclaw.py

```diff
@@ -0,0 +1,1451 @@
+"""
+This file is part of CLIMADA.
+
+Copyright (C) 2017 ETH Zurich, CLIMADA contributors listed in AUTHORS.
+
+CLIMADA is free software: you can redistribute it and/or modify it under the
+terms of the GNU Lesser General Public License as published by the Free
+Software Foundation, version 3.
+
+CLIMADA is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License along
+with CLIMADA. If not, see <https://www.gnu.org/licenses/>.
+
+---
+
+Inundation from TC storm surges, modeled using the library GeoClaw
+
+This module requires a Fortran compiler (such as gfortran) to run!
+"""
+
+import __main__
+import contextlib
+import datetime as dt
+import importlib
+import logging
+import re
+import pathlib
+import site
+import subprocess
+import sys
+import warnings
+
+import cartopy.crs as ccrs
+import cartopy.feature as cfeature
+from cartopy.mpl.gridliner import LONGITUDE_FORMATTER, LATITUDE_FORMATTER
+from matplotlib.backends.backend_agg import FigureCanvasAgg
+import matplotlib.colors
+import matplotlib.cm
+import matplotlib.pyplot as plt
+from mpl_toolkits.axes_grid1.inset_locator import inset_axes
+import numpy as np
+import pandas as pd
+import scipy.sparse as sp
+import xarray as xr
+
+from climada import CONFIG
+from climada.hazard import Centroids, Hazard, Tag as TagHazard, TropCyclone
+from climada.hazard.tc_tracks import estimate_rmw, estimate_roci
+from climada.util import ureg
+import climada.util.coordinates as u_coord
+
+LOGGER = logging.getLogger(__name__)
+
+HAZ_TYPE = 'TCSurgeGeoClaw'
+"""Hazard type acronym for this module."""
+
+CLAWPACK_GIT_URL = CONFIG.hazard.tc_surge_geoclaw.resources.clawpack_git.str()
+"""URL of the official Clawpack git repository."""
+
+CLAWPACK_VERSION = CONFIG.hazard.tc_surge_geoclaw.resources.clawpack_version.str()
+"""Version or git decorator (tag, branch) of Clawpack to use."""
+
+CLAWPACK_SRC_DIR = CONFIG.hazard.tc_surge_geoclaw.clawpack_src_dir.dir()
+"""Directory for Clawpack source checkouts (if it doesn't exist)"""
+
+GEOCLAW_WORK_DIR = CONFIG.hazard.tc_surge_geoclaw.geoclaw_work_dir.dir()
+"""Base directory for GeoClaw run data."""
+
+KN_TO_MS = (1.0 * ureg.knot).to(ureg.meter / ureg.second).magnitude
+NM_TO_KM = (1.0 * ureg.nautical_mile).to(ureg.kilometer).magnitude
+MBAR_TO_PA = (1.0 * ureg.mbar).to(ureg.pascal).magnitude
+KM_TO_DEG = 1.0 / (60 * NM_TO_KM)
```

Comment:
I was confused about this also in the other review of wildfire. Why do a conversion between degrees and km with a constant? The ratio depends on the coordinate system right? Also, the longitude conversion depends on the latitude value...

Reply:
That constant is perfectly fine: It stores the degree distance at the equator of one kilometer. That's a constant, so it makes sense to store it. When actually doing a conversion from km to degrees, you would do the following:
```
lat = y * KM_TO_DEG
lon = x * KM_TO_DEG * np.cos(lat)
```
However, if you don't care about a factor 2 and you are in a TC context, then you can even decide to ignore the cosine factor. For example, in case of this module, the factor is used to give a rough estimate of the size of the coastal area that is actually threatened to be flooded.

Reply:
Could we maybe change the name of the variable to `KM_TO_DEG_EQUA` or something similar?

Reply:
Done.